### ISSUE TITLE
(Bloom) Unexported source identifier field cause incorrect gap calculation leading to OSS blocks data lost

### ISSUE THREAD

**Issue by honganan** [grafana/loki#13115]
**Describe the bug**
`exportTSInSecs` field in TSDB source identifier not exported. 

When we running bloom-compactor with a TSDB file which timestamp in filename is unix `Seconds` format, the Meta generated will always do not contains this field after serielized and uploaded to OSS, which will default to false when we read it. 

During the next interval even if no TSDB changed, it will still calculate a gap in `gapsBetweenTSDBsAndMetas()` in `controller.go`. And running a empty gap will generate none block but a new meta, and then delete the old metas and blocks. That result the OSS blocks data lost.

**To Reproduce**
1. Having a TSDB who's timestamp in filename is unix seconds format, e.g., 1717378350-compactor-1717284265317-1717375930531-1dfd963d.tsdb.gz
2. Stop data ingestion to keep TSDB no changes and start Loki bloom-compactor (main branch)
3. After it finish and wait for next interval and watch the log
4. It still calculate a gap the second time and it run nothing but delete the old blocks and meta

**Expected behavior**
bloom-compactor should calculate no gap as no TSDB changed the second time

**Environment:**
 - Infrastructure: [e.g., Kubernetes, bare-metal, laptop]
 - Deployment tool: [e.g., helm, jsonnet]

**Screenshots, Promtail config, or terminal output**
If applicable, add any output to help explain your problem.

**Comment by honganan** [comment]
Additionally, Should we have a check if `buildGaps()` really build blocks before check outdated metas and delete OSS data?

For now if gap calculate have any problem will cause OSS data lost.

**Comment by owen-d** [comment]
Hey, I'm having trouble understanding some of this.

* The `exportTSInSecs` is set in `ParseSingleTenantTSDBPath` which is in the call-path of where the bloom-compactors load TSDBs. It sets this field based on the number of digits in the timestamps. Therefore, I expect this to be present.

That being said, I can add a separate check at the end to ensure we don't delete blocks that are also present in the "new" compaction cycle -- what do you think?

**Comment by honganan** [comment]
> * The `exportTSInSecs` is set in `ParseSingleTenantTSDBPath` which is in the call-path of where the bloom-compactors load TSDBs. It sets this field based on the number of digits in the timestamps. Therefore, I expect this to be present.

Yes this part works fine, The problem is `Meta.Sources` we created in any compaction cycle doesn't contains this field, and We need to compare `Meta.Sources[i].name()` with TSDB's `name()` when calculate gaps in `controller.go#gapsBetweenTSDBsAndMetas()` 

https://github.com/grafana/loki/blob/1d6f8d51fcfd1c806159e17bce978ea39ee5936c/pkg/bloomcompactor/controller.go#L666
It will not be equal here, this leads to the calculation of a gap that should not be there.

**Comment by chaudum** [comment]
hi @honganan 

I still don't understand the problem, tbh.
The field `exportTSInSecs` does not need to be exported in order to be used in an internal struct function such as `str()`, which is called by `s.Name()`.

**Comment by honganan** [comment]
Imagine the following process:
Suppose we have a TSDB index file that needs to generate to bloom: 1717378350-compactor-1717284265317-1717375930531-1dfd963d.tsdb.gz

In the first round:
We download this TSDB file and parse the filename to determine that exportTSInSecs is true. Then, we generate bloom blocks and meta from this file and upload them to S3. Imagine what the exportTSInSecs in Meta.Source would be here (it would be false because this field is not exported and therefore cannot be serialized into Meta.Source).

In the second round:
We download this TSDB file and the newly generated Meta file, and calculate whether there is a gap in `controller.go#gapsBetweenTSDBsAndMetas()`. One of the steps in calculating whether there is a gap is comparing:
TSDB.Name() with Meta.Source.Name() 

Then, the problem occurs. Meta.Source.Name() has now become a nanosecond format, similar to: 1717378350000000000-compactor-1717284265317-1717375930531-1dfd963d.tsdb.gz. This Meta will be considered unrelated to this TSDB file.

**Comment by honganan** [comment]
In summary, the `Meta`(now in loki/pkg/storage/stores/shipper/bloomshipper/client.go) structure contains a Sources []tsdb.SingleTenantTSDBIdentifier field, and within tsdb.SingleTenantTSDBIdentifier, there is an exportTSInSecs field. When Meta is serialized and written to OSS storage, the exportTSInSecs field is being set to its default value, false, which is not the expected behavior. It should retain its actual value; otherwise, this will lead to issues during subsequent gap calculations.

https://github.com/grafana/loki/blob/b7fe6bf63d9c64f4c7d4636ac6d470d4ca918517/pkg/storage/stores/shipper/bloomshipper/client.go#L104
